source/dub/dub.d

```diff
@@ -275,16 +275,77 @@ class Dub {
 
 		const dubFolderPath = NativePath(thisExePath).parentPath;
 
+		// override default userSettings + localRepository if a $DPATH or
+		// $DUB_HOME environment variable is set.
+		bool overrideUserSettings;
+		bool overrideLocalRepository;
+		{
+			string dpathOverride = environment.get("DUB_HOME");
+			if (!dpathOverride.length) {
+				dpathOverride = environment.get("DPATH");
+				if (dpathOverride.length)
+					dpathOverride = (NativePath(dpathOverride) ~ "dub/").toNativeString();
+
+			}
+			if (dpathOverride.length) {
+				overrideUserSettings = true;
+				overrideLocalRepository = true;
+
+				m_dirs.userSettings = NativePath(dpathOverride);
+				m_dirs.localRepository = m_dirs.userSettings;
+			}
+		}
+
+		// load system-wide configs:
 		readSettingsFile(m_dirs.systemSettings ~ "settings.json");
 		readSettingsFile(dubFolderPath ~ "../etc/dub/settings.json");
 		version (Posix) {
 			if (dubFolderPath.absolute && dubFolderPath.startsWith(NativePath("usr")))
 				readSettingsFile(NativePath("/etc/dub/settings.json"));
 		}
+
+		// Override user + local package path from /etc/dub/settings.json
+		// Then continues loading local settings from these folders. (keeping
+		// global /etc/dub/settings.json settings intact)
+		//
+		// Don't use it if either $DPATH or $DUB_HOME are set.
+		if (!overrideUserSettings) {
+			if (this.userSettingsOverride.length) {
+				m_dirs.userSettings = NativePath(this.userSettingsOverride.expandEnvironmentVariables);
+
+				overrideUserSettings = true;
+			} else if (this.dubHome.length) {
+				m_dirs.userSettings = NativePath(this.dubHome.expandEnvironmentVariables);
+
+				overrideUserSettings = true;
+			}
+		}
```

Comment:
I'm a bit lost... Why do we have `dubHome` AND `userSettings` ?
Also use `this.m_config.userSettings.set` for condition, instead of checking length.